File: api-service/app/schemas/video.py
"""Video API schema definitions."""

from datetime import datetime

from pydantic import BaseModel, ConfigDict


class VideoSourceResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    source_code: str
    display_name: str
    video_type: str
    provider: str
    channel_or_stream_id: str | None = None
    embed_url: str | None = None
    playback_url: str | None = None
    thumbnail_url: str | None = None
    title: str | None = None
    description: str | None = None
    region: str | None = None
    country: str | None = None
    city: str | None = None
    topic_tags: list[str]
    license_mode: str
    priority: int
    enabled: bool
    rollout_state: str
    status: str
    notes: str | None = None
    source_metadata: dict


class VideoSourcePatchRequest(BaseModel):
    display_name: str | None = None
    embed_url: str | None = None
    playback_url: str | None = None
    thumbnail_url: str | None = None
    title: str | None = None
    description: str | None = None
    region: str | None = None
    country: str | None = None
    city: str | None = None
    topic_tags: list[str] | None = None
    license_mode: str | None = None
    priority: int | None = None
    enabled: bool | None = None
    rollout_state: str | None = None
    status: str | None = None
    notes: str | None = None
    source_metadata: dict | None = None


class VideoProbeCheckpointResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    source_code: str
    job_code: str
    last_probe_at: datetime | None = None
    last_success_at: datetime | None = None
    last_http_status: int | None = None
    last_error: str | None = None
    is_live: bool
    last_title: str | None = None
    last_thumbnail: str | None = None
    consecutive_failures: int


class VideoSourceWithHealthResponse(VideoSourceResponse):
    checkpoint: VideoProbeCheckpointResponse | None = None


class VideoHealthResponse(BaseModel):
    sources: list[VideoSourceWithHealthResponse]


class VideoProbeResponse(BaseModel):
    ok: bool
    source_code: str
    job_code: str
    status: str
    is_live: bool
    http_status: int | None = None
    title: str | None = None
    thumbnail_url: str | None = None
    error: str | None = None

File: web-app/src/types/video.ts
export type VideoType = 'youtube_embed' | 'hls';
export type VideoRolloutState = 'draft' | 'poc' | 'canary' | 'default' | 'paused';
export type VideoStatus = 'unknown' | 'live' | 'offline' | 'error';

export interface VideoProbeCheckpoint {
  id: string;
  source_code: string;
  job_code: string;
  last_probe_at?: string | null;
  last_success_at?: string | null;
  last_http_status?: number | null;
  last_error?: string | null;
  is_live: boolean;
  last_title?: string | null;
  last_thumbnail?: string | null;
  consecutive_failures: number;
}

export interface VideoSource {
  id: string;
  source_code: string;
  display_name: string;
  video_type: VideoType;
  provider: string;
  channel_or_stream_id?: string | null;
  embed_url?: string | null;
  playback_url?: string | null;
  thumbnail_url?: string | null;
  title?: string | null;
  description?: string | null;
  region?: string | null;
  country?: string | null;
  city?: string | null;
  topic_tags: string[];
  license_mode: string;
  priority: number;
  enabled: boolean;
  rollout_state: VideoRolloutState;
  status: VideoStatus;
  notes?: string | null;
  source_metadata: Record<string, unknown>;
  checkpoint?: VideoProbeCheckpoint | null;
}

export interface VideoHealthResponse {
  sources: VideoSource[];
}

export interface VideoProbeResponse {
  ok: boolean;
  source_code: string;
  job_code: string;
  status: VideoStatus;
  is_live: boolean;
  http_status?: number | null;
  title?: string | null;
  thumbnail_url?: string | null;
  error?: string | null;
}
